### PR TITLE
[autoscaler]: Kill workers if the monitor raises an exception

### PR DESCRIPTION
**Pull request by ls-daniel**
For example, if Redis dies on the head node, we don't want to end up with
100 expensive nodes sitting around burning resources.

We keep the head node up in order to have some logs to look at.

### CODE REVIEW

**Review by markgoodhead — APPROVED** [review]

**Review by ericl — COMMENTED** [review]
Hmm, a lot of the times I want to keep the workers up when doing interactive debugging. Perhaps only scale down to min cluster size? That way, it would more accurately simulate an idle job.

**Comment by ls-daniel** [comment]
Well, the issue here is that in the event that the autoscaler has died, our internal state might not be reliable.

I suppose we could pull the list of all workers, select N of them randomly, and kill all except those N?

**Comment by ls-daniel** [comment]
This commit changes the behaviour such that the cluster will kill all but min_workers worker nodes.

The nodes are selected at random. We could use LoadMetrics data but at this point, the autoscaler has crashed for whatever reason, and I think we want to avoid hunting around in potentially outdated/invalid state.

File: python/ray/monitor.py [hunk 1]
```diff
@@ -325,10 +328,39 @@ def run(self):
...
+                    override_cluster_name=None,
+                    keep_min_workers=True,  # Retain minimal amount of workers.
+                )
+                clean = True
+                logger.info("Monitor: Workers taken down.")
+            except:
+                logger.error("Monitor: Cleanup exception. Trying again...")
+                time.sleep(2)
+
+        return
```
Comment (ericl) [inline python/ray/monitor.py]: Don't need this return

**Review by ericl — APPROVED** [review]

File: python/ray/scripts/scripts.py [hunk 2]
```diff
@@ -487,6 +487,11 @@ def create_or_update(cluster_config_file, min_workers, max_workers, no_restart,
     is_flag=True,
     default=False,
     help="Only destroy the workers.")
+@click.option(
+    "--keep-min-workers",
+    is_flag=True,
+    default=False,
+    help="Retain the minimal amount of workers specified in the config.")
```
Comment (ericl) [inline python/ray/scripts/scripts.py]: Consider not exposing this to the command line

**Comment by edoakes** [comment]
@ericl should we fix conflicts and merge this or close?

**Comment by markgoodhead** [comment]
@ericl Can I ask why this was closed? We're happy to fix the merge issues if that's a blocker - this is a very useful safety feature for us and would have saved a lot of cost from autoscaler crashes we've had where we've silently left up expensive AWS machines for a long time from this.

**Comment by ericl** [comment]
I bulk closed all stale pull requests -- feel free to reopen if still
needed.

**Comment by markgoodhead** [comment]
I don't think I have access to re-open but we're happy to take a look at fixing the merge conflicts next month so this can be merged if you're still OK with it being included?

**Comment by ericl** [comment]
Reopened

**Comment by CJosephides** [comment]
I've fixed the conflicts. Could you please take a look, @ericl ?

**Comment by markgoodhead** [comment]
@ericl @edoakes Would one of you be able to get this into master now we've fixed the conflicts?

**Pull request merged**